File: GridView/RadGridViewColumn-Indicate-Editor-Type/RadGridViewColumn-indicate-editor-type-VB/IndicatedBrowseColumn/IndicatedBrowseCellElement.vb
Imports Telerik.WinControls
Imports Telerik.WinControls.UI

Public Class IndicatedBrowseCellElement
    Inherits GridBrowseCellElement

    Private indicator As BrowseEditorButton

    Public Sub New(ByVal column As GridViewColumn, ByVal row As GridRowElement)
        MyBase.New(column, row)
        indicator = New BrowseEditorButton()

        'this code adds a registration for RadDropDownListArrowButtonElement in order to allow its usage in other controls
        Dim theme As Theme = ThemeRepository.ControlDefault
        Dim sg As StyleGroup = theme.FindStyleGroup("Telerik.WinControls.UI.RadBrowseEditor")
        sg.Registrations.Add(New StyleRegistration("Telerik.WinControls.UI.BrowseEditorButton"))

        indicator.Alignment = ContentAlignment.MiddleRight
        indicator.ShouldHandleMouseInput = True
        indicator.NotifyParentOnMouseInput = False

        indicator.FitToSizeMode = Telerik.WinControls.RadFitToSizeMode.FitToParentBounds
        indicator.MaxSize = New System.Drawing.Size(30, 15)
        indicator.Margin = New System.Windows.Forms.Padding(0, 1, 4, 1)
        AddHandler indicator.Click, AddressOf indicator_Click

        Me.Children.Add(indicator)

    End Sub

    Protected Overrides Sub OnCellFormatting(ByVal e As CellFormattingEventArgs)
        MyBase.OnCellFormatting(e)
        If indicator IsNot Nothing Then
            indicator.Visibility = If(CType(Me.ColumnInfo, IndicatedBrowseColumn).EnableIndicator = True, ElementVisibility.Visible, ElementVisibility.Collapsed)
        End If
    End Sub

    Private Sub indicator_Click(ByVal sender As Object, ByVal e As EventArgs)
        AddHandler Me.GridControl.CellEditorInitialized, AddressOf grid_CellEditorInitialized
        Me.GridControl.EndEdit()
        Me.GridControl.BeginEdit()
    End Sub

    Private Sub grid_CellEditorInitialized(ByVal sender As Object, ByVal e As GridViewCellEventArgs)

        Dim currentEditor As GridBrowseEditor = TryCast(e.ActiveEditor, GridBrowseEditor)
        If currentEditor IsNot Nothing Then
            RemoveHandler Me.GridControl.CellEditorInitialized, AddressOf grid_CellEditorInitialized

            Dim element As RadBrowseEditorElement = TryCast(currentEditor.EditorElement, RadBrowseEditorElement)
            element.BrowseButton.PerformClick()
        End If
    End Sub

    Protected Overrides ReadOnly Property ThemeEffectiveType() As Type
        Get
            Return GetType(GridBrowseCellElement)
        End Get
    End Property

    Public Overrides Function IsCompatible(ByVal data As GridViewColumn, ByVal context As Object) As Boolean
        Return TypeOf data Is IndicatedBrowseColumn AndAlso TypeOf context Is GridDataRowElement
    End Function
End Class

File: GridView/GridViewIndicatedColumns/GridViewCustomCellsC#/IndicatedBrowseColumn/IndicatedBrowseCellElement.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Text;
using Telerik.WinControls;
using Telerik.WinControls.UI;

namespace GridViewCustomCells
{
    public class IndicatedBrowseCellElement : GridBrowseCellElement
    {
        private BrowseEditorButton indicator;

        public IndicatedBrowseCellElement(GridViewColumn column, GridRowElement row) : base(column, row)
        {
            indicator = new BrowseEditorButton();

            //this code adds a registration for RadDropDownListArrowButtonElement in order to allow its usage in other controls
            Theme theme = ThemeRepository.ControlDefault;
            StyleGroup sg = theme.FindStyleGroup("Telerik.WinControls.UI.RadBrowseEditor");
            sg.Registrations.Add(new StyleRegistration("Telerik.WinControls.UI.BrowseEditorButton"));

            indicator.Alignment = ContentAlignment.MiddleRight;
            indicator.ShouldHandleMouseInput = true;
            indicator.NotifyParentOnMouseInput = false;

            indicator.FitToSizeMode = Telerik.WinControls.RadFitToSizeMode.FitToParentBounds;
            indicator.MaxSize = new System.Drawing.Size(30, 15);
            indicator.Margin = new System.Windows.Forms.Padding(0, 1, 4, 1);
            indicator.Click += indicator_Click;

            this.Children.Add(indicator);
            
        }
       
        protected override void OnCellFormatting(CellFormattingEventArgs e)
        {
            base.OnCellFormatting(e);
            if (indicator != null)
            {
                indicator.Visibility = ((IndicatedBrowseColumn)this.ColumnInfo).EnableIndicator == true ? ElementVisibility.Visible : ElementVisibility.Collapsed;
            }
        }

        void indicator_Click(object sender, EventArgs e)
        {
            this.GridControl.CellEditorInitialized += grid_CellEditorInitialized;
            this.GridControl.EndEdit();
            this.GridControl.BeginEdit();
        }

        void grid_CellEditorInitialized(object sender, GridViewCellEventArgs e)
        {
            GridBrowseEditor editor = e.ActiveEditor as GridBrowseEditor;
            if (editor != null)
            {
                this.GridControl.CellEditorInitialized -= grid_CellEditorInitialized;

                GridBrowseEditorElement element = editor.EditorElement as GridBrowseEditorElement;
                element.BrowseButton.PerformClick();
            }
        }

        protected override Type ThemeEffectiveType
        {
            get
            {
                return typeof(GridBrowseCellElement);
            }
        }

        public override bool IsCompatible(GridViewColumn data, object context)
        {
            return data is IndicatedBrowseColumn && context is GridDataRowElement;
        }
    }
}